[REFACTOR] : 뮤지컬 등록 리팩토링

Seongjun-Kwon opened the pull request:
## PR Description 🗒️
뮤지컬 등록 API 리팩토링

close #72 

## 추가 사항 및 설명 ➕
- MusicalService 가 너무 많은 Repository 에 의존하고 많은 책임을 가지게 되어 facade 레이어를 추가하였습니다.
- 아직 S3 에 저장하게 구현하지는 못했고 저장할 path 를 지정할 수 있게는 해놨습니다~

## 수정 사항 및 이유 🔄
- ViewRating 과 Genre 는 String 이 아니기에 @NotBlank 가 아닌 @NotNull 로 수정했습니다.

## To Reviewers 🙏
리뷰해야할 파일이 많은데 공연 등록이 연관되어 있는 엔티티가 많아서 어쩔 수가 없네요.
꼼꼼한 리뷰 부탁드려요~

Seongjun-Kwon committed:
[MOVE] : SeatGrade 파일 위치 이동

- SeatGrade 도 엔티티이므로 독립적인 패키지를 가지게 하였음.

Seongjun-Kwon committed:
[FEAT] : 뮤지컬 등록을 위해 필요한 Repository 생성

Seongjun-Kwon committed:
[FEAT] : 뮤지컬 등록을 위해 필요한 DTO 구현

Seongjun-Kwon committed:
[FEAT] : 뮤지컬 등록 API 구현

Seongjun-Kwon committed:
[REFACTOR] : 뮤지컬 썸네일을 저장하는 테이블 분리

Seongjun-Kwon committed:
[REFACTOR] : 뮤지컬 등록 API 리팩토링

- 기존에 MusicalService 가 여러 Repository 에 의존하던 것을 Facade 레이어를 만들어서 해결함

Seongjun-Kwon committed:
[CHORE] : origin/feature/#62 브랜치로부터 pull

Seongjun-Kwon committed:
[RENAME] : DTO 컨벤션에 맞춰서 파일 이름 수정

Seongjun-Kwon committed:
[MOVE] : musical_detail_image 패키지 안의 파일들 musical 패키지로 합침

- MusicalDetailImage 는 다른 엔티티들과 연관되어있지 않고 오직 Musical 에만 종속되기 때문

twotwobread commented:
이 PR 닫고 로그인/회원가입하는 브랜치로 다시 PR 날리겠습니다!!

twotwobread commented:
어 내 PR인줄 알고 잘못닫았어요 ㅠㅠ 미안해요 ㅠㅠ

park-se-jun commented on src/main/java/com/prgrms/be/intermark/domain/musical/dto/MusicalCreateRequestDTO.java:
```diff
@@ -0,0 +1,43 @@
+	@NotNull @Positive int runningTime,
+	@NotNull long managerId,
+	@NotNull long stadiumId,
+	List<Long> actors,
```
actors 면 actor 객체들로 생각 될 수도 있을 것 같아서 actorIds 로 하는거 어떤가요?

Seongjun-Kwon replied:
좋네요. 반영했습니다~

park-se-jun commented on src/main/java/com/prgrms/be/intermark/domain/actor/service/ActorService.java:
```diff
@@ -0,0 +1,29 @@
+	private final ActorRepository actorRepository;
+
+	@Transactional(readOnly = true)
+	public List<Actor> findActors(List<Long> actorIds) {
```
이 메소드가 사용되는 위치 역시 service 레이어인데, 파사드 서비스에서 findById를 for문으로 돌리지 않고, 전체를 내려서 내부에서 진행하는 이유가 따로 있을까요?

Seongjun-Kwon replied:
기존에 MusicalService 의 create 메서드 내에 뮤지컬을 생성하고, 공연장을 조회하고, 좌석 등급을 만들고, 뮤지컬_좌석을 만드는 등...(생략) 많은 로직을 함께 두었습니다.
테스트를 해본다고 생각했을 때, MusicalService 가 9개의 Repository 에 의존하고 있기에 단위테스트를 하기가 어렵다고 생각했습니다. 추가적으로 create 메서드가  너무 많은 책임을 가지고 있기도 하고요.
그래서 기존에 MusicalService 의 create 메서드가 가지고 있던 공연장을 조회하고, 좌석 등급을 만들고....(생략) 등의 각각의 로직은 각각의 서비스들이 갖게 하고 그것들을 호출하여 조립하는 것을 MusicalFacadeService 에서 하면 좋겠다는 생각을 했습니다.

세준님이 말씀하신대로 Facade 서비스에서 findById 를 for 문으로 돌리면 Actor 리스트를 받는 책임을 Facade 가 가지게 되므로 ActorService 내부에서 처리하고 Facade 는 그것을 호출만하는 방식을 선택하였습니다.

여기서 추가로 고민된 부분은 Actor Id 리스트를 받아서 Actor 리스트를 찾는 로직이 Actor 관련 api 에서 사용할 일이 있을까하는 것이었습니다.
현재 리뷰해주신 버전의 코드는 사용할 일이 있다고 판단하고 작성한 것인데, 댓글 작성하면서 생각해보니 CastingService 외에는 사용할 일이 없을 것 같아서 CastingService 안으로 넣어주었습니다.

park-se-jun commented on src/main/java/com/prgrms/be/intermark/domain/casting/service/CastingService.java:
```diff
@@ -0,0 +1,31 @@
+	private final CastingRepository castingRepository;
+
+	@Transactional
+	public void saveCasting(List<Actor> actors, Musical musical) {
```
이거도 리스트를 받고 내부에서 처리한 이유가 궁금합니당

Seongjun-Kwon replied:
위에 남겼습니다.

park-se-jun commented on src/main/java/com/prgrms/be/intermark/domain/user/dto/UserFindResponseDTO.java:
```diff
@@ -0,0 +1,14 @@
+import lombok.Builder;
+
+@Builder
+public record UserFindResponseDTO() {
```
저희를 생각해서 형식만 만들어 둔건가요?

Seongjun-Kwon replied:
로직에 필요해서 만들려다가 필요없어졌는데 삭제를 안했네요.
삭제했습니다.

park-se-jun commented on src/main/java/com/prgrms/be/intermark/domain/stadium/dto/StadiumFindResponseDTO.java:
```diff
@@ -0,0 +1,17 @@
+
+	public static StadiumFindResponseDTO from(Stadium stadium) {
+		return StadiumFindResponseDTO.builder()
+			.name(stadium.getAddress())
```
name 에 address가 들어가나요?

Seongjun-Kwon replied:
반영했습니다~

park-se-jun commented on src/main/java/com/prgrms/be/intermark/domain/musical/controller/MusicalController.java:
```diff
@@ -1,27 +1,54 @@
-    @GetMapping
-    public ResponseEntity<PageResponseDTO<Musical, MusicalResponseDTO>> getAllMusicals(Pageable pageable) {
-        PageResponseDTO<Musical, MusicalResponseDTO> musicals = musicalService.findAllMusicals(pageable);
+	@GetMapping
```
url 이 빠져 있는거 같아요

Seongjun-Kwon replied:
반영했습니다

park-se-jun commented on src/main/java/com/prgrms/be/intermark/domain/musical/dto/MusicalCommandResponseDTO.java:
```diff
@@ -0,0 +1,15 @@
+import lombok.Builder;
+
+@Builder
+public record MusicalCommandResponseDTO(Long id) {
```
어디에 쓰이는 건가요?

Seongjun-Kwon replied:
등록, 수정 같은 커맨드 시 응답을 위한 DTO 입니다.

Seongjun-Kwon committed:
[TEST] : 파일명, 패키지 변경사항 테스트에도 반영

park-se-jun commented on src/main/java/com/prgrms/be/intermark/domain/musical/model/Musical.java:
```diff
@@ -64,10 +74,16 @@ public class Musical {
     @Column(name = "running_time", nullable = false)
     private int runningTime;
 
+    @NotNull
```
궁금한게 원래도 DetailImage랑 썸네일이랑은 별개이면서, 썸네일 URL은 바로 공연에 저장하고, DetailImage만 별개의 테이블이었던거 아닌가요? pr 메시지 보고 제가 잘못 이해하고 있었던 건가 싶어서 남겨용~~

Seongjun-Kwon replied:
기존 기획대로 수정했습니다~

park-se-jun commented on src/main/java/com/prgrms/be/intermark/domain/musical/model/MusicalDetailImage.java:
```diff
@@ -21,6 +30,10 @@ public class MusicalDetailImage {
     @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
 
+    @NotBlank
+    @Column(name="original_file_name", nullable = false)
```
이거가 필요한 필드인지 궁금함니당

Seongjun-Kwon replied:
필수적이지는 않은 것 같아요
다만, 보통 게시글 같은 곳에 첨부파일 등록할 때, 제가 등록했던 파일 이름으로 저장되있던 것이 떠올라서 넣었습니다.

park-se-jun commented on src/main/java/com/prgrms/be/intermark/domain/musical/model/MusicalThumbnail.java:
```diff
@@ -0,0 +1,39 @@
+
+	@NotBlank
+	@Column(name="original_file_name", nullable = false)
+	private String originalFileName;
```
이거두 필요한 건지 궁금하다는 느낌쓰

Seongjun-Kwon replied:
위와 같습니다

park-se-jun reviewed (commented):
정말 고생 많으셨습니당,다른거 다 merge 되고 merge 해야 된다며 앞으로 할일이 아직 많이 남아있지 않나.... 성준이형 화이팅~~~ 질문이 조금 있어서 comment로다가 남길게요!!

twotwobread commented on src/main/java/com/prgrms/be/intermark/domain/musical/dto/MusicalSeatCreateRequestDTO.java:
```diff
@@ -0,0 +1,11 @@
+
+@Builder
+public record MusicalSeatCreateRequestDTO(
+	@NotBlank Long seatId,
```
Long 타입인데 @NotBlank를 단 이유가 있을까요??

Seongjun-Kwon replied:
ㄷㄷ 수정했습니다

twotwobread commented on src/main/java/com/prgrms/be/intermark/domain/seatgrade/repository/SeatGradeRepository.java:
```diff
@@ -0,0 +1,13 @@
+
+public interface SeatGradeRepository extends JpaRepository<SeatGrade, Long> {
+
+	Optional<SeatGrade> findSeatGradeByNameAndMusical(String name, Musical musical);
```
쿼리 메소드 사용할 때 엔티티명을 생략하면 조금 더 어떤 필드를 이용해서 조회하는지 한눈에 잘 보일 것 같습니다.

Seongjun-Kwon replied:
감사합니다~

twotwobread reviewed (approved):
수고하셨습니다!

Seongjun-Kwon committed:
[RENAME] : 직관적으로 필드 명 변경

Seongjun-Kwon committed:
[DEL] : 사용하지 않는 UserFindResponseDTO 삭제

Seongjun-Kwon committed:
[FIX] : 잘못 작성한 코드 수정

- MusicalController 의 getAllMusicals 메서드에 매핑 url 실수
- StadiumFindResponseDTO 의 name 필드에 address 매핑해준 실수
- MusicalSeatCreateRequestDTO 의 seatId 필드에 NotBlank 제약조건 준 실수

Seongjun-Kwon committed:
[RENAME] : 메서드 명 읽기 쉽게 변경

Seongjun-Kwon committed:
[REFACTOR] : UploadImageService 로 이미지 업로드 기능 추상화

rhdtn311 reviewed (approved):
고생하셨습니다!

Seongjun-Kwon committed:
[REFACTOR] : ActorService 안에 구현했던 findActors 를 CastingService 안에 넣어서 응집도 높힘

Seongjun-Kwon committed:
[CHORE] : origin/feature/#62 브랜치로부터 pull
